### PR TITLE
Nadro/3716/mvp revolve

### LINKED ISSUE #3716: MVP Revolve UI
MVP can be just the same as extrude, where we automatically pick the "X" axis for you (we can merge this but hide the button etc unless in dev mode)

We're going to be modeling this off of extrude since there are some similarities.

When the user clicks the extrude button, the experience is large interacting with the command pallet, that's why when you look at where the extrude button is defined in `src/lib/toolbar.ts` the Xstate action it fires off is `Find and select command`

```ts
{
  id: 'extrude',
  onClick: ({ commandBarSend }) =>
    commandBarSend({
      type: 'Find and select command',
      data: { name: 'Extrude', groupId: 'modeling' },
    }),
  disabled: (state) => !state.can('Extrude'),
  icon: 'extrude',
  status: 'available',
  title: 'Extrude',
  hotkey: 'E',
  description: 'Pull a sketch into 3D along its normal or perpendicular.',
  links: [{ label: 'KCL docs', url: 'https://zoo.dev/docs/kcl/extrude' }],
},
```

To make something similar work for revolve you'll need to add a config to `src/lib/commandBarConfigs/modelingCommandConfig.ts` because this is the MVP we can make this exactly the same as Extrude and automatically use `axis: "X"` so we don't need to add a selection for the axis for the time being, since there's plenty to work on in getting the code-mods working.

Once the cmd pallet cmd is added, you should be able to get it to fire an event in the main modeling app that should now have all of the details needed for the code mod
- Selection of the solid2d/sketch
- how far to revolve
- and we're going to assume the axis is just "X" for now

You will need to create a new event in the modeling state machine that triggers a `AST revolve` action. Again we can copy what Extrude is already doing.
image
An event from `idle` going back to `idle` with our action, just with revolve names instead of extrude.

Here's the typing for the extrude event

image

image

So we're pretty much just going to want something like
```ts
Revolve: {
    selection: Selections
    angle: KclCommandValue
    // TODO axis
  }
```

Then when it comes to writing the AST mod
We can pretty much copy [AST extrude](https://github.com/KittyCAD/modeling-app/blob/c2e9d18f92cb911cc1f5787af432d597adaa4eb3/src/machines/modelingMachine.ts#L1086)

The difference should just be the specifics

The core of the extrude though is this here

```ts
const extrudeCall = createCallExpressionStdLib('extrude', [
    distance,
    shouldPipe
      ? createPipeSubstitution()
      : createIdentifier(variableDeclarator.id.name),
  ])
```
We can ignore the shouldPipe so it's really
```ts
const extrudeCall = createCallExpressionStdLib('extrude', [
    distance,
    createIdentifier(variableDeclarator.id.name),
])
```
so for revolve it will something like
```ts
const revolveCall = createCallExpressionStdLib('revolve', [
    createObjectExpression({
      angle: angle,
      axis: createLiteral('X'),
    }),
    createIdentifier(variableDeclarator.id.name),
  ])
```

Hopefully that make sense why it look a little different with the `createObjectExpression` because extrude looks like
`const extrude01 = extrude(20, sketch002)`
And revolve looks like
`const revolve01 = revolve({ angle: 90, axis: "X" }, sketch001)`

I think I'll leave it there, I'm probably missing stuff that we'll have to figure out along the way.

One final tip is, if you want to look at the ast is to use the ast explorer.

https://github.com/user-attachments/assets/e5b6d02d-1d7f-43db-b25d-e22c9d4d0b92

### PR DESCRIPTION
**Pull request by nadr0**
closes #3716 

----

Accidentally branched off a previous  branch but that is merged into main now. It is okay, the commit history just has those commits.

Feature starts with commit dd2a53f44f10128cb3288f4da76dd2c162497a00

-----

I implemented a copied workflow from Extrude to implement Revolve. I renamed all but one set of functions from extrude to revolve when doing the implementation.

I left a few `TODO` comments to indicate that the revolve is in MVP and was referenced from extrude. More work to come that is outlined in https://github.com/KittyCAD/modeling-app/issues/3714

You can see a video [here](https://github.com/KittyCAD/modeling-app/issues/3716#issuecomment-2319072803)

Going through the revolve workflow from the button in the toolbar will give an output like this
Screenshot from 2024-08-30 11-39-24

### CODE REVIEW

**Comment by jessfraz** [comment]
needs maybe unit tests for the ast mods and playwright to make sure we dont break it

**Comment by Irev-Dev** [comment]
Because of
image
it really doesn't touch other code so I'm not worried about breaking other things (plus existing tests should cover that)

Because this is incomplete (hard codes revolve axis to 'X') I'm a little skeptical that it's worth writing tests at this stage since they will have to change anyway, but I think worth merging allows us to have lots of concise PRs. (it's disabled unless your in dev so users won't see it)

Is this ready for review btw @nadr0 ?

**Comment by Irev-Dev** [comment]
hmm when I wrote the above I was thinking more e2e tests, unit tests are definitely less onerous and will have to change less. But having said that they will still change significantly, because later it will involve the selection of the revolve axis. I'm in two minds about it.

**Review by Irev-Dev — APPROVED** [review]
Nice, my comments are small and mostly related to existing issues.

There seems to be some problems with code migrated to v5 still, but you demoed it working for me so maybe you still have a commit you haven't push?

File: src/lang/modifyAst.ts [hunk 1]
```diff
@@ -342,6 +342,103 @@ export function extrudeSketch(
   }
 }
 
+export function revolveSketch(
+  node: Program,
+  pathToNode: PathToNode,
+  shouldPipe = false,
+  angle = createLiteral(4) as Expr
```
[the remark this answers is not in the document]
Comment (nadr0) [reply src/lang/modifyAst.ts]: Updated, thanks for the suggestion. I'll keep that in mind when I see any `as`.

File: src/lang/modifyAst.ts [hunk 2]
```diff
@@ -342,6 +342,103 @@ export function extrudeSketch(
...
+):
+  | {
+      modifiedAst: Program
+      pathToNode: PathToNode
+      pathToRevolveArg: PathToNode
+    }
+  | Error {
+  const _node = { ...node }
```
[the remark this answers is not in the document]
Comment (nadr0) [reply src/lang/modifyAst.ts]: Updated this for revolve and extrude.

File: src/machines/modelingMachine.ts [hunk 3]
```diff
@@ -566,6 +572,57 @@ export const modelingMachine = setup({
...
+        const { selection, angle } = event.data
+        let ast = kclManager.ast
+        if (
+          'variableName' in angle &&
+          angle.variableName &&
+          angle.insertIndex !== undefined
+        ) {
+          const newBody = [...ast.body]
+          newBody.splice(
+            distance.insertIndex,
```
[the remark this answers is not in the document]
Comment (nadr0) [reply src/machines/modelingMachine.ts:588]: I believe it should be `angle`. I missed that when I renamed `distance` to `angle`. 

This pattern is the same one in `AST extrude`.

File: src/lang/modifyAst.ts [hunk 4]
```diff
@@ -342,6 +342,103 @@ export function extrudeSketch(
...
+  // We're not creating a pipe expression,
+  // but rather a separate constant for the extrusion
+  const name = findUniqueName(node, KCL_DEFAULT_CONSTANT_PREFIXES.REVOLVE)
+  const VariableDeclaration = createVariableDeclaration(name, revolveCall)
+
+  const sketchIndexInPathToNode =
+    pathToDecleration.findIndex((a) => a[0] === 'body') + 1
+  const sketchIndexInBody = pathToDecleration[
+    sketchIndexInPathToNode
+  ][0] as number
```
[the remark this answers is not in the document]
Comment (nadr0) [reply src/lang/modifyAst.ts]: Updated.

**Comment by nadr0** [comment]
> There seems to be some problems with code migrated to v5 still, but you demoed it working for me so maybe you still have a commit you haven't pushed?

Everything should be pushed.

**Pull request merged**